libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
 	if (!_literal.annotation().type)
 		m_errorReporter.fatalTypeError(_literal.location(), "Invalid literal value.");
 }
+
+void ConstantEvaluator::endVisit(Identifier const& _identifier)
+{
+	VariableDeclaration const *variableDeclaration = dynamic_cast<VariableDeclaration const *>(_identifier.annotation().referencedDeclaration);
+	if (!variableDeclaration)
+		return;
+	if (!variableDeclaration->isConstant())
+	{
+		m_errorReporter.fatalTypeError(_identifier.location(), "Identifier must be declared constant.");
+	}
+	else
+	{
+		ASTPointer<Expression> value = variableDeclaration->value();
+		if (value)
+		{
+			if (!value->annotation().type)
+				ConstantEvaluator e(*value, m_errorReporter);
```

Comment:
Are we sure this cannot lead to infinite recursion at this point in the compiler? Also, is the recursion depth limited?

Reply:
For the second point, it is fine to just add a counter in ConstantEvaluator and fold if it gets too high.

Reply:
@chriseth Can't think of a scenario that can lead to infinite recursion. It is a good idea to have a recursion depth limit though. What depth would be a good one? 10? 20? How do we decide? And what error to throw in that case?

Reply:
@chriseth You are correct, a simple "uint constant LEN = LEN;" caused an infinite recursion (and crashed my laptop :( ). Will add a recursion limit of 20 (?).

Reply:
Done.